src/lsm/manifest_level.zig

```diff
@@ -54,6 +55,78 @@ pub fn ManifestLevelType(
             tables: std.BoundedArray(TableInfoReference, constants.lsm_growth_factor),
         };
 
+        pub const LevelKeyRange = struct {
```

Comment:
How did you compute the 38MB? That is a lot higher than my estimate. :thinking:

Reply:
Oops, I ended up using _table_count_max_for_tree_ (`2396744`) for this, mistook it for the number of keys in the worst case. But actually, doesn't this mean that the worst case is this number multiplied by the `value_count_max` we compute in `state_machine.zig`? I seem to be missing basic something here. 😅

Reply:
- I don't think this has any relation to `value_count_max`.
- We only need to enable this data structure on id and object trees. The other trees, if I understand correctly, don't use the key range.
- Each entry in the `tables_key_min` corresponds to a single table. (16 bytes or 8 bytes, respectively).
- The LSM is configured for a maximum of ~2.3M tables total. But those 2.3M tables are distributed dynamically between the different trees -- the secondary index trees (which is most of them) have empty `tables_key_min` segmentedarrays.

Reply:
Argh, brain fog! I did use _table_count_max_for_tree_ correctly then, since there is a single entry per table in `tables_key_min`. 

I thought that the main advantage of checking the ManifestLevel range in `iterator_start` 
was that lookups into _any tree_ (including our `IndexTree`s) could benefit from our key range, such that levels that absolutely can't contain the key can be eliminated using the range: https://github.com/tigerbeetle/tigerbeetle/blob/0ef3ccd19368f9d0f96558d79924d8a791ac9b26/src/lsm/manifest_level.zig#L415-L424

However, I haven't yet observed any code that performs lookup into IndexTrees. 🤔

Reply:
> But those 2.3M tables are distributed dynamically between the different trees

I think I missed spotting this, are we enforcing this invariant somewhere? 

My understanding was that ~2.3M was the count for the maximum number of tables _per tree_, as opposed to the entire forest:
https://github.com/tigerbeetle/tigerbeetle/blob/0ef3ccd19368f9d0f96558d79924d8a791ac9b26/src/lsm/tree.zig#L55-L59

Reply:
> I think I missed spotting this, are we enforcing this invariant somewhere?

It's not properly enforced yet throughout LSM, but it is referenced in the superblock manifest.

2.3M is the maximum number of tables per tree, but also the maximum for the whole forest.
That is, at the LSM level, there are at most 2.3M tables total, and in the extreme case they may all belong to a single tree.
(In practice, though, that distribution is impossible with our state machine.)

https://github.com/tigerbeetle/tigerbeetle/blob/1e1b685bb74abdceeae7f207580f2ed23f926447/src/vsr/superblock_manifest.zig#L55-L57

> However, I haven't yet observed any code that performs lookup into IndexTrees.

It isn't implemented yet; Rafael is working on it.

Reply:
### For posterity

See https://github.com/tigerbeetle/tigerbeetle/pull/944#issuecomment-1642816324 for resolution.